### PR TITLE
Prioritizing of virtual columns in hive partitioning

### PR DESCRIPTION
**Pull request by yariks5s**
### Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
...

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Run only fuzzers related jobs (libFuzzer fuzzers, AST fuzzers, etc.)
- [ ]  Exclude: AST fuzzers
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll4** [comment]
*This is an automated comment for commit 24eeaffa7a3ddbfa0fb7bc4546942bc18cab06af with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/68606/24eeaffa7a3ddbfa0fb7bc4546942bc18cab06af/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/f323c76431147134c987c16d7fea52abddb0895f/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

**Review by KochetovNicolai — APPROVED** [review]

File: src/Storages/VirtualColumnUtils.cpp [hunk 1]
```diff
@@ -129,38 +129,43 @@ NameSet getVirtualNamesForFileLikeStorage()
...
+        if (storage_columns.has(name) && !context->getSettingsRef().use_hive_partitioning)
             return;
+        if (storage_columns.has(name))
+        {
+            local_type = storage_columns.get(name).type;
+            storage_columns.remove(name);
+        }
 
-        desc.addEphemeral(name, type, "");
+        desc.addEphemeral(name, local_type, "");
```
Comment (Avogar) [inline src/Storages/VirtualColumnUtils.cpp:168]: We do the same check `storage_columns.has(name)` twice when `use_hive_partitioning=1`. Let's avoid it.

File: src/Storages/VirtualColumnUtils.cpp [hunk 2]
```diff
@@ -129,38 +129,43 @@ NameSet getVirtualNamesForFileLikeStorage()
...
     auto add_virtual = [&](const auto & name, const auto & type)
     {
-        if (storage_columns.has(name))
+        auto local_type = type;
+        if (storage_columns.has(name) && !context->getSettingsRef().use_hive_partitioning)
             return;
+        if (storage_columns.has(name))
+        {
+            local_type = storage_columns.get(name).type;
+            storage_columns.remove(name);
```
Comment (Avogar) [inline src/Storages/VirtualColumnUtils.cpp:165]: We need tests for it

File: src/Storages/VirtualColumnUtils.cpp [hunk 3]
```diff
@@ -129,36 +129,45 @@ NameSet getVirtualNamesForFileLikeStorage()
...
 
     auto add_virtual = [&](const auto & name, const auto & type)
     {
         if (storage_columns.has(name))
+        {
+            if (!context->getSettingsRef().use_hive_partitioning)
+                return;
+
+            if (storage_columns.size() == 1)
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Cannot implement partition by all columns in a file");
```
Comment (Avogar) [inline src/Storages/VirtualColumnUtils.cpp:165]:
```suggestion
                throw Exception(ErrorCodes::INCORRECT_DATA, "Cannot use hive partitioning for file {}: it contains only partition columns. Disable use_hive_partitioning setting to read this file", path);
```

**Review by Avogar — APPROVED** [review]

**Pull request merged**